pkg/host/common.go

```diff
@@ -0,0 +1,267 @@
+}
+
+// NewTopic creates a new topic
+func (n *hostImpl) NewTopic(name string, opts ...ps.TopicOpt) (*ps.Topic, *ps.TopicEventHandler, *ps.Subscription, error) {
+	// Check if PubSub is Set
+	if n.PubSub == nil {
+		return nil, nil, nil, errors.New("NewTopic: Pubsub has not been set on SNRHost")
+	}
+
+	// Call Underlying Pubsub to Connect
+	t, err := n.Join(name, opts...)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Create Event Handler
+	h, err := t.EventHandler()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Create Subscriber
+	s, err := t.Subscribe()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return t, h, s, nil
+}
+
+// Router returns the host node Peer Routing Function
+func (hn *hostImpl) Router(h host.Host) (routing.PeerRouting, error) {
+	// Create DHT
+	kdht, err := dht.New(hn.ctx, h)
+	if err != nil {
+		hn.SetStatus(Status_FAIL)
+		return nil, err
+	}
+
+	// Set Properties
+	hn.IpfsDHT = kdht
+
+	// Setup Properties
+	return hn.IpfsDHT, nil
+}
+
+// PubSub returns the host node PubSub Function
+func (hn *hostImpl) Pubsub() *ps.PubSub {
+	return hn.PubSub
+}
+
+// Routing returns the host node Peer Routing Function
+func (hn *hostImpl) Routing() routing.Routing {
+	return hn.IpfsDHT
+}
+
+// SetStreamHandler sets the handler for a given protocol
+func (n *hostImpl) SetStreamHandler(protocol protocol.ID, handler network.StreamHandler) {
+	n.host.SetStreamHandler(protocol, handler)
+}
+
+// SendMessage writes a protobuf go data object to a network stream
+func (h *hostImpl) SendMessage(id peer.ID, p protocol.ID, data proto.Message) error {
+	err := h.HasRouting()
+	if err != nil {
+		return err
+	}
+
+	s, err := h.NewStream(h.ctx, id, p)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	// marshall data to protobufs3 binary format
+	bin, err := proto.Marshal(data)
+	if err != nil {
+		return err
+	}
+
+	// Create Writer and write data to stream
+	w := msgio.NewWriter(s)
+	if err := w.WriteMsg(bin); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Stat returns the host stat info
+func (hn *hostImpl) Stat() (map[string]string, error) {
+	// Return Host Stat
+	return map[string]string{
+		"ID":        hn.host.ID().String(),
+		"Status":    hn.status.String(),
+		"MultiAddr": hn.host.Addrs()[0].String(),
+	}, nil
+}
+
+// Serve handles incoming peer Addr Info
+func (hn *hostImpl) Serve() {
+	for {
+		select {
+		case mdnsPI := <-hn.mdnsPeerChan:
+			if err := hn.Connect(mdnsPI); err != nil {
+				hn.host.Peerstore().ClearAddrs(mdnsPI.ID)
+				continue
+			}
+
+		case dhtPI := <-hn.dhtPeerChan:
+			if err := hn.Connect(dhtPI); err != nil {
+				hn.host.Peerstore().ClearAddrs(dhtPI.ID)
+				continue
+			}
+		case <-hn.ctx.Done():
+			return
+		}
+	}
+}
+
+// VerifyData verifies incoming p2p message data integrity
+func (n *hostImpl) VerifyData(data []byte, signature []byte, peerId peer.ID, pubKeyData []byte) bool {
+	key, err := crypto.UnmarshalPublicKey(pubKeyData)
```

Comment:
Does this lock us into a single key format? it doesnt seem like we will be able to support atribitrary key types (unless this is just for wallets) this seems incompatible with did verification method key types